### PR TITLE
[#19] Implement Attribution model

### PR DESCRIPTION
**Pull request by tessi**

### CODE REVIEW

**Comment by klappradla** [comment]
_quickly rebasing on Tessi's branch again_

File: models/attribution.js [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+}
+
+function t(lang, key) {
+  return ATTRIBUTION_TRANSLATIONS[lang][key];
+}
+function isStringPresent(string) {
+  return typeof string === 'string' && string.length > 0;
+}
+
+function validateParams({
```
Comment (klappradla) [inline models/attribution.js]: 🤔 I feel like we should introduce a wrapper object for `fileData`... Here you're expecting `fileUrl`, and `fileTitle` which are referred to as `rawUrl` and `title` in other services...

If we'd have a common wrapper we could simply assert that the thing we're receiving has this interface...

Background: figuring out how to rename the attributes when passing them in from other services is what's currently the most work for me when taking over #15 ...

Comment (tessi) [reply models/attribution.js]: ✅  i agree, implemented that change and commited it to this PR

File: .tool-versions [hunk 2]
```diff
@@ -0,0 +1 @@
+nodejs 10.14.2
```
[the remark this answers is not in the document]
Comment (tessi) [reply .tool-versions]: urgh, didn't mean to commit this. you're right (and i already have this set to legacy version files for that reason) ❤️

File: models/attribution.js [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    self.fileTitle
+  )}“${editingAttribution}${urlSnippet}`;
+}
+
+class Attribution {
+  constructor(params) {
+    validateParams(params);
+    Object.assign(this, params);
+
+    this.fileTitle = params.fileInfo.title;
```
Comment (klappradla) [inline models/attribution.js]: Can we not just move this up and then assign all things at once? E.g. like so (only rough sketch):
```javascript
validateParams(params);
const { title: fileTitle, rawUrl: fileUrl } = params.fileInfo;
// do the HTML things...
Object.assign(this, {...params, fileTitle, fileUrl, artistHtml, attributionHtml });

File: models/attribution.js [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    if (isStringPresent(this.artistHtml)) {
+      this.artistHtml = sanitizeHtml(this.artistHtml);
+    }
+    if (isStringPresent(this.attributionHtml)) {
+      this.attributionHtml = sanitizeHtml(this.attributionHtml);
+    }
+  }
+
+  html() {
+    if (this.typeOfUse === 'offline') {
```
Comment (klappradla) [inline models/attribution.js]: Would spare the if for things where we return immediately:
```javascript
html() {
  const { typeOfUse } = this;
  return typeOfUse === 'offline' ? getPrint... : getHTML...;
 }
```

File: models/attribution.js [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+ * and refactored only slightly to make comparisons between this app and the
+ * original Lizenhinweisgenerator easier.
+ */
+
+const assert = require('assert');
+const jsdom = require('jsdom');
+const License = require('./license');
+const HtmlSaniziter = require('../services/htmlSanitizer');
+
+const { JSDOM } = jsdom;
```
[the remark this answers is not in the document]
Comment (tessi) [reply models/attribution.js]: this is how jsdom recommended it on its website
https://github.com/jsdom/jsdom#basic-usage

[the remark this answers is not in the document]
Comment (klappradla) [reply models/attribution.js]: Sound good!

**Comment by tessi** [comment]
Hey @klappradla thanks for your throughout review! I applied some proposals (extracting translations, the constants thing) but (as we already talked about in person) would leave the object-deconstruction and single-line-if-statements for a later PR. I agree with your opinion, especially when private methods still accept `self` as the first param but desconstruct `self` into used variables in the first line of a method. I would leave this to a follow-up cleansing PR because of time pressure on other matters.

File: models/attribution.test.js [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+    it('asserts valid modification', () => {
+      expect(() => newAttribution({ modification: 123 })).toThrow();
+    });
+
+    it('asserts valid modificationAuthor', () => {
+      expect(() => newAttribution({ modificationAuthor: 123 })).toThrow();
+    });
+  });
+
+  describe('when initialized with defaults', () => {
```
Comment (klappradla) [inline models/attribution.test.js]: How about doing all of these as snapshots? Does not clutter the file that much...

File: models/attribution.test.js [hunk 7]
```diff
@@ -0,0 +1,316 @@
+const Attribution = require('./attribution');
+const License = require('./license');
+
+describe('attribution', () => {
+  const exampleCC4License = new License({
```
Comment (klappradla) [inline models/attribution.test.js]: How about moving all of these into a factory? Would declutter this file a bit.

**Comment by tessi** [comment]
@klappradla I created a [ticket](https://ora.pm/project/59434/kanban/task/745932) to follow up on your comments when there is more time. I would love to come back to it when follow-up PRs/tickets are not as pressing.

**Pull request merged**